[SPARK-11316] coalesce doesn't handle UnionRDD with partial locality properly

tgravescs opened the pull request:
## What changes were proposed in this pull request?

coalesce doesn't handle UnionRDD with partial locality properly.  I had a user who had a UnionRDD that was made up of mapPartitionRDD without preferred locations and a checkpointedRDD with preferred locations (getting from hdfs).  It took the driver over 20 minutes to setup the groups and put the partitions into those groups before it even started any tasks.  Even perhaps worse is it didn't end up with the number of partitions he was asking for because it didn't put a partition in each of the groups properly.

The changes in this patch get rid of a n^2 while loop that was causing the 20 minutes, it properly distributes the partitions to have at least one per group, and it changes from using the rotation iterator which got the preferred locations many times to get all the preferred locations once up front.  

Note that the n^2 while loop that I removed in setupGroups took so long because all of the partitions with preferred locations were already assigned to group, so it basically looped through every single one and wasn't ever able to assign it.  At the time I had 960 partitions with preferred locations and 1020 without and did the outer while loop 319 times because that is the # of groups left to create.  Note that each of those times through the inner while loop is going off to hdfs to get the block locations, so this is extremely inefficient.
## How was the this patch tested?

Added unit tests for this case and ran existing ones that applied to make sure no regressions. 
Also manually tested on the users production job to make sure it fixed their issue.  It created the proper number of partitions and now it takes about 6 seconds rather then 20 minutes.
 I did also run some basic manual tests with spark-shell doing coalesced to smaller number, same number, and then greater with shuffle.

Commit:
[SPARK-11316] coalesce setupGroups doesn't handle UnionRDD with partial localtiy properly

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51784/
Test FAILed.

tgravescs commented:
Jenkins, test this please

SparkQA commented:
**[Test build #51786 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51786/consoleFull)** for PR 11327 at commit [`afe14dc`](https://github.com/apache/spark/commit/afe14dce508b1e51820f16e33f09c9aa402bca3e).

SparkQA commented:
**[Test build #51786 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51786/consoleFull)** for PR 11327 at commit [`afe14dc`](https://github.com/apache/spark/commit/afe14dce508b1e51820f16e33f09c9aa402bca3e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51786/
Test FAILed.

Commit:
fix scalastyle

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51787/
Test FAILed.

tgravescs commented:
Jenkins, test this please

SparkQA commented:
**[Test build #51789 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51789/consoleFull)** for PR 11327 at commit [`c9eb032`](https://github.com/apache/spark/commit/c9eb032af8e453a5ba6776279cf0cd6946d0cd55).

SparkQA commented:
**[Test build #51789 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51789/consoleFull)** for PR 11327 at commit [`c9eb032`](https://github.com/apache/spark/commit/c9eb032af8e453a5ba6776279cf0cd6946d0cd55).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51789/
Test FAILed.

Commit:
remove extra whitespace

SparkQA commented:
**[Test build #51799 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51799/consoleFull)** for PR 11327 at commit [`8665114`](https://github.com/apache/spark/commit/86651146e90d5126d379a4abc6d73a8c6b7a50df).

SparkQA commented:
**[Test build #51799 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51799/consoleFull)** for PR 11327 at commit [`8665114`](https://github.com/apache/spark/commit/86651146e90d5126d379a4abc6d73a8c6b7a50df).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51799/
Test PASSed.

rxin commented:
cc @davies

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -192,7 +192,8 @@ private class PartitionCoalescer(maxPartitions: Int, prev: RDD[_], balanceSlack:
         prev.partitions.iterator.flatMap(p => {
-          if (currPrefLocs(p).size > x) Some((currPrefLocs(p)(x), p)) else None
+          val locs = currPrefLocs(p)
+          if (locs.size > x) Some((locs(x), p)) else None
```
locs.size -> locs.length

`size` could be O(N), while `length` is O(1)

tgravescs replied:
Are you sure on this?  Overall I'm fine with changing but just want to understand for future.

locs here is a Seq

From scaladoc:
The size of this sequence, equivalent to length

scala doc says length on sequence also says:
Note: will not terminate for infinite-sized collections.

Looking at the scala source code:
https://github.com/scala/scala/blob/2.10.x/src/library/scala/collection/SeqLike.scala#L106

davies replied:
cc @rxin

rxin replied:
It's not size vs length. It's seq.size can sometimes be O(n).

For size vs length, we should use length if it is a string or an array, but size if it is a collection.

tgravescs replied:
@rxin I don't follow your first sentence? 

Second sentence says use size for a collection and Seq is a collection.

rxin replied:
I was just responding to @davies' comment. Not saying anything wrong with this.

tgravescs commented:
Are there any other comments about functionality?

tgravescs commented:
ping @davies  was there any other concern or does this look good?

davies commented:
@tgravescs I did not have enough time to look into the details yet (not familar this part), sorry for the delay.

tgravescs commented:
ping @davies @rxin   Any chance I can get review on this?

davies commented:
@tgravescs I'm reviewing this now, sorry for the late.

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -324,6 +319,40 @@ private class PartitionCoalescer(maxPartitions: Int, prev: RDD[_], balanceSlack:
+      }
+      // we have went through all with preferred locations now just make sure one
+      // partition per group
+      val numEmptyPartitionGroups = groupArr.length - getPartitions.length
```
Should we move this into setupGroups(), so we still meet the assumption that all the groups have one partition?

tgravescs replied:
i split it this way because one is called setupGroups and the throwballs one is put stuff in the groups. I see the fact we put stuff in them in setupgroup as an optimization rather then a necessity. 

Do you see a benefit to do it there vs here?

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -324,6 +319,40 @@ private class PartitionCoalescer(maxPartitions: Int, prev: RDD[_], balanceSlack:
+      val numEmptyPartitionGroups = groupArr.length - getPartitions.length
+      val partitionsNotInGroups = prev.partitions.filter(p => !initialHash.contains(p))
+      for (i <- 0 until math.min(numEmptyPartitionGroups, partitionsNotInGroups.length)) {
+        groupArr.find(pg => pg.size == 0).map(firstEmpty => {
```
This is still O(N*N)

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -324,6 +319,40 @@ private class PartitionCoalescer(maxPartitions: Int, prev: RDD[_], balanceSlack:
+          // handle both cases and its not easy to get # of partitions with preferred locs
+          var (nxt_replica, nxt_part) = rotIt.next()
+          if (!initialHash.contains(nxt_part)) {
+            groupArr.find(pg => pg.size == 0).map(firstEmpty => {
```
This is also O(N*N)

davies commented:
@tgravescs The current change is good for your case, I'm thinking that maybe we could do better.

The LocationIterator has a bad smell, it may call getPreferredLocs() many times on the same partition, which could be expensive as you mentioned, we should only call getPreferredLocs() on a partition once, by caching the result of it, or get rid of LocationIterator totally.

Since we will call getPreferredLocs on every partition of previous RDD, we could eaglely call them in the beginning, partition all the partitions into two groups: with prefered location and without, do different things on them. This approach should solve all the cases that 1) all partition have locs 2) some partitions have locs 3) none of partiions have locs.

Hopefully, this could be done in less than 10 seconds in your case.

tgravescs commented:
thanks for the feedback.  I'm fine with that change and actually had considered it, I just wasn't sure if the intention of the location iterator was to handle the locations changing (in some case I'm not aware of) so I was going the less invasive method of leaving that part the same.

If we aren't aware of any issues with that I'll make the changes

davies commented:
I think the current implementation does not handle location changing, and we can't.

Commit:
Rework to remove the rotation iterator to minimize calls to get preferred locations

Commit:
Merge remote-tracking branch 'upstream/master' into SPARK-11316

Commit:
Rework to upmerge to latest

tgravescs commented:
Ok, I replaced the location iterator and now get all the preferred locations up front.  This made the run time of the this go from around a minute down to around 6 seconds. 

I kept most of the logic the same and just changed how its getting the locations.

SparkQA commented:
**[Test build #57550 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57550/consoleFull)** for PR 11327 at commit [`2eff583`](https://github.com/apache/spark/commit/2eff583d896b1032477a299aa9ae488711d5f01c).

SparkQA commented:
**[Test build #57550 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57550/consoleFull)** for PR 11327 at commit [`2eff583`](https://github.com/apache/spark/commit/2eff583d896b1032477a299aa9ae488711d5f01c).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class PartitionLocations(prev: RDD[_]) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57550/
Test FAILed.

davies commented:
@tgravescs That's great, could you fix the style?

Commit:
fix style

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -169,43 +169,41 @@ private class DefaultPartitionCoalescer(val balanceSlack: Double = 0.10)
+        partsWithLocs.toArray.flatMap(parts => {
+          val p = parts._1
+          val locs = parts._2
+          if (locs.size > x) Some((locs(x), p)) else None
```
We may just append (locs(x), p) to an ArrayBuffer

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -169,43 +169,41 @@ private class DefaultPartitionCoalescer(val balanceSlack: Double = 0.10)
-      }
-    }
+  // gets the *current* preferred locations from the DAGScheduler (as opposed to the static ones)
+  def currPrefLocs(part: Partition, prev: RDD[_]): Seq[String] = {
```
private or inline this?

SparkQA commented:
**[Test build #57556 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57556/consoleFull)** for PR 11327 at commit [`f012cd5`](https://github.com/apache/spark/commit/f012cd5fc20feb20088c808275cf283d0f594cec).

SparkQA commented:
**[Test build #57556 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57556/consoleFull)** for PR 11327 at commit [`f012cd5`](https://github.com/apache/spark/commit/f012cd5fc20feb20088c808275cf283d0f594cec).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57556/
Test FAILed.

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -289,10 +284,12 @@ private class DefaultPartitionCoalescer(val balanceSlack: Double = 0.10)
    */
-  def pickBin(p: Partition, prev: RDD[_], balanceSlack: Double): PartitionGroup = {
+  def pickBin(p: Partition, prev: RDD[_], balanceSlack: Double,
+      partitionLocs: PartitionLocations): PartitionGroup = {
```
```
def pickBin(
    p: Partition,
    prev: RDD[_],
    balanceSlack: Double,
    partitionLocs: PartitionLocations): PartitionGroup = {
```

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -320,7 +317,8 @@ private class DefaultPartitionCoalescer(val balanceSlack: Double = 0.10)
 
-  def throwBalls(maxPartitions: Int, prev: RDD[_], balanceSlack: Double) {
+  def throwBalls(maxPartitions: Int, prev: RDD[_],
+      balanceSlack: Double, partitionLocs: PartitionLocations) {
```
indents

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -334,8 +332,41 @@ private class DefaultPartitionCoalescer(val balanceSlack: Double = 0.10)
+      while (partNoLocIter.hasNext && initialHash.size < groupArr.size) {
+        var nxt_part = partNoLocIter.next()
+        if (!initialHash.contains(nxt_part)) {
+          groupArr.find(pg => pg.numPartitions == 0).map(firstEmpty => {
```
This is still O(N*N) (the worst), it could be
```
groupArr.filter(pg => pg.numPartitions == 0).foreach { pg =>
  while (partNoLocIter && pg.numPartitions == 0) {
     if (!initialHash.contains(nxt_part)) {
       pg.partitions += nxt_part
         initialHash += nxt_part
     }
  }
}
```

davies commented on core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala:
```diff
@@ -334,8 +332,41 @@ private class DefaultPartitionCoalescer(val balanceSlack: Double = 0.10)
+        while (partIter.hasNext && initialHash.size < groupArr.size) {
+          var (nxt_replica, nxt_part) = partIter.next()
+          if (!initialHash.contains(nxt_part)) {
+            groupArr.find(pg => pg.numPartitions == 0).map(firstEmpty => {
```
same here

davies commented:
@tgravescs Could you also update the description of reflect the new changes?

Commit:
fix comments from review

tgravescs commented:
thanks for the review, made changes and updated description.

SparkQA commented:
**[Test build #57644 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57644/consoleFull)** for PR 11327 at commit [`305a7db`](https://github.com/apache/spark/commit/305a7db60a2fc836035ed06c8207ce772c5e3b23).

davies commented:
LGTM, pending tests. It's great to have 200X speedup, thanks!

SparkQA commented:
**[Test build #57644 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57644/consoleFull)** for PR 11327 at commit [`305a7db`](https://github.com/apache/spark/commit/305a7db60a2fc836035ed06c8207ce772c5e3b23).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57644/
Test FAILed.

SparkQA commented:
**[Test build #2965 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2965/consoleFull)** for PR 11327 at commit [`305a7db`](https://github.com/apache/spark/commit/305a7db60a2fc836035ed06c8207ce772c5e3b23).

tgravescs commented:
test failure is in ExternalAppendOnlyMapSuite which is unrelated. I'll kick jenkins again.

tgravescs commented:
Jenkins, test this please

SparkQA commented:
**[Test build #57653 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57653/consoleFull)** for PR 11327 at commit [`305a7db`](https://github.com/apache/spark/commit/305a7db60a2fc836035ed06c8207ce772c5e3b23).

SparkQA commented:
**[Test build #2965 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2965/consoleFull)** for PR 11327 at commit [`305a7db`](https://github.com/apache/spark/commit/305a7db60a2fc836035ed06c8207ce772c5e3b23).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #57653 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57653/consoleFull)** for PR 11327 at commit [`305a7db`](https://github.com/apache/spark/commit/305a7db60a2fc836035ed06c8207ce772c5e3b23).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57653/
Test FAILed.

tgravescs commented:
Test build #2965 has finished successfully so I'm going by that.  The other one had another unit test failure that is unrelated.

davies commented:
Merging this into master, thanks!
